### DISCUSSION TITLE
Proposal: Credentials Helper Support in Reclient

### DISCUSSION THREAD

**Discussion by gkousik** [bazelbuild/reclient#16] [Ideas]
# Background

Reclient has the following authentication mechanisms to talk to a corresponding RBE server:

1. Application default credentials (RBE\_use\_application\_default\_credentials)
2. GCE Credentials (RBE\_use\_gce\_credentials)
3. External Authentication token (RBE\_use\_external\_auth\_token)

Methods (1) and (2) provide support for authenticating with an RBE server implementation running on GCP and method (3) is sufficient to authenticate with RBE servers that expect an “Authorization: Bearer &lt;token>” token. While in method (3) the user can specify that they want to provide a custom token, there is no clear option in reclient for them to actually specify a token ([creds\_file](https://github.com/bazelbuild/reclient/blob/main/cmd/reproxy/main.go#L128) can be used to specify the token, but its not clear how reproxy can refresh the credentials in this file).

In addition, there have been use cases where users want to provide custom HTTP headers for auth. Given this, the proposal in this doc is to implement support for a credentials helper interface in reclient.

# Credentials Helper

A credentials helper would be a binary that can be provided to reclient. Reclient would call the binary to receive a set of headers to append to RBE RPCs along with information about when those headers expire.

## Interface

Two new flags will be introduced:

### Invoking the credentials helper

```
-creds_helper      - specifies the path to the credentials helper binary
-creds_helper_args - specifies the arguments to pass when invoking the
                     credentials helper binary
```

### Return value of credentials helper

The credentials helper binary will be expected to return the following response.

```
{
    "headers": {
        "AuthorizationHeader_1": ["Value 1", "Value 2"],
        "AuthorizationHeader_2": ["Value 1", "Value 2"]
    },
    "expiry": "<timestamp>",
    "refresh_expiry": "<timestamp>"
}
```

Note that this is very similar to the interface of the Bazel credentials helper with two optional return options:

1. “expiry”            - This will be a timestamp that indicates when the authorization headers expire. Reproxy will call the credentials helper to get new headers when the expiry is past. If no expiry is specified, the tokens in the headers are considered to NOT have an expiration.
2. “refreshExpiry” - This will be a timestamp that indicates the time until which reproxy can continue to call the credentials helper programmatically, without the user having to re authenticate themselves. \
This is useful in cases where the credentials helper relies on an authentication mechanism that requires the user to periodically reauthenticate themselves. If no refreshExpiry is specified, it is assumed to be infinite. \
This field is intended to only be used by the bootstrap binary. When bootstrap sees refresh\_expiry to be valid, it would assume that reproxy would be able to obtain valid authorization headers from the credentials helper and continue with reproxy startup. Otherwise, bootstrap would fail the reproxy startup process and indicate to the user that they will need to reauthenticate. \
 \
For both expiry and refreshExpiry, “YYYY-MM-DD HH:MM:SS.000 &lt;TIMEZONE>” format will be used.

#### RBE RPCs

Once the right authorization headers are returned from reclient, there needs to be a way to insert these headers into the RPCs being sent to RBE. With gRPC, a custom TokenSource can be created to insert custom headers per RPC request. Here’s a short example:

```
type customTokenSource struct {}

func (s customTokenSource) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
	hdrs := make(map[string]string)
	hdrs["Authorization"] = "Bearer token" // Insert headers from creds helper here.
	return hdrs, nil
}

func (s customTokenSource) RequireTransportSecurity() bool {
	return true
}

// In reproxy
func main() {
   …
   		clientOpts = append(clientOpts, &client.PerRPCCreds{Creds: &customTokenSource{}})
		go func() {
			log.Infof("Creating a new SDK client")
			grpcClient, err := rflags.NewClientFromFlags(initCtx, clientOpts...)
    …
}
```

### Credentials Helper Flag Location

There are two places where the creds helper flags can be defined:

#### Option (1) - In reclient (i.e., in bootstrap and reproxy binaries)

In this option, both [bootstrap](https://github.com/bazelbuild/reclient/blob/main/cmd/bootstrap/main.go) and reproxy binaries (and any other binary that relies on the credentials helper like [remotetool](https://github.com/bazelbuild/remote-apis-sdks/tree/master/go/cmd/remotetool)) will define this flag. When it is set, bootstrap / reproxy would call the credentials helper to get the headers, then inject them into RBE RPCs using the mechanism described in [this section](#header-injection).

Cons

1. Multiple definitions of the same flags across binaries.
2. In the future, any binary wanting to use the credentials helper to talk to RBE will also have to redefine these flags (e.g., remotetool in remote-apis-sdks repository).

#### Option (2) - In remote-apis-sdks repository

In this option, the flags will be defined in the remote-apis-sdks repository alongside the [other authorization flags](https://github.com/bazelbuild/remote-apis-sdks/blob/master/go/pkg/flags/flags.go#L30). When the creds helper flag is specified, the SDK will do the following:

1. Initialize the Client with a new TokenSource as described below. Initially, there won’t be any headers in the token source. 
2. The first time this token source is asked for Request Metadata through “GetRequestMetadata()” call, there will be a blocking call to call the credentials helper and obtain the tokens. The obtained tokens will then be stored in memory along with the expiry.
3. Subsequent calls to GetRequestMetadata() within the expiration period will return the tokens stored in memory.
4. Whenever the headers are near expiration, they will be refreshed through a call to the credentials helper.

```shell
type credsHelperTokenSource struct {
    headers map[string]string
    headersLock sync.RWMutex
    expiry time.Time

    credsHelperPath string
    credsHelperArgs []string
}

func (s *credsHelperTokenSource) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
    return updateToken()
}

func (s *credsHelperTokenSource) RequireTransportSecurity() bool {
    return true
}

func (s *credsHelperTokenSource) token(credsHelperPath string, credsHelperArgs []string) error {
    s.headersLock.Lock()
    defer s.headersLock.Unlock()

    if expiry.Sub(time.Now()) <= 5 * time.Minute {
        return s.headers, nil
    }
    ...
    // call creds helper to obtain new headers, save it to headers map.
    return s.headers, nil
}
```

Even with this approach, [bootstrap](https://github.com/bazelbuild/reclient/blob/main/cmd/bootstrap/main.go) binary in reclient would still need to define the credentials helper flags since it doesn’t initialize the SDK and will need to be able to say if the specified credentials helper can be used to obtain valid credentials or not to fail the build early.

**Proposed solution:** Pick Option (2) but move the auth related flags to a separate auth package within the remote-apis-sdks. This auth package can then be imported by the SDK and other clients binaries like [bootstrap](https://github.com/bazelbuild/reclient/blob/main/cmd/bootstrap/main.go) that don’t necessarily initialize the RBE client in remote-apis-sdks.

**Comment by Yannic** [comment]
LGTM in general. Would you mind opening an issue in github.com/EngFlow/credential-helper-spec for the new fields though please so we can standardize them? I‘d like to avoid having slightly different flavours of the protocol 😅

**Comment by Yannic** [comment]
We've recently standardized expiration of credentials in https://github.com/EngFlow/credential-helper-spec/pull/2

**Comment by magreenblatt** [comment]
This also appears to be a blocker for reclient usage with BuildBuddy. https://www.buildbuddy.io/docs/guide-auth/

**Comment by magreenblatt** [comment]
Has some portion of this proposal already been implemented? It looks like an external `credentials_helper` binary is currently supported (added mid-2024), but usage and scope of support is unclear. Thanks.

EDIT: Looks like the credentials helper might already support this Proposal. Code [here](https://github.com/bazelbuild/reclient/blob/3e4c533c42b323356abaf11d2050435e5ab2d7d9/cmd/reproxy/main.go#L276), example usage [here](https://chromium-review.googlesource.com/c/chromium/tools/build/+/5447758).

Can anyone provide an example of using an external `credentials_helper` to supply a custom HTTP header for auth? Is there a specific JSON response required by NewExternalCredentials to return a [PerRPCCredentials](https://pkg.go.dev/google.golang.org/grpc/credentials#PerRPCCredentials) object?

EDIT2: Found the [docs](https://pkg.go.dev/github.com/bazelbuild/remote-apis-sdks/go/pkg/credshelper#NewExternalCredentials) for NewExternalCredentials which mostly match the JSON example in this Proposal. I'll give that a try.

**Comment by magreenblatt** [comment]
I've verified that BuildBuddy RBE authentication (custom HTTP header) works using the EngFlow Chromium [setup](https://blog.engflow.com/2024/02/20/how-to-build-chromium-with-reclient/) (see "Sanity check: verify that Reclient works" section) with minor changes.

Environment variables:
```
export RBE_service=[mydomain].buildbuddy.io:443
export RBE_automatic_auth=false
export RBE_use_gce_credentials=false
export RBE_use_application_default_credentials=false
export RBE_credentials_helper=/path/to/cred-helper.sh
export RBE_credentials_helper_args="arg1 arg2"
```

Example contents of `cred-helper.sh`:
```
#!/bin/bash

# https://www.buildbuddy.io/docs/guide-auth/
apikey="YOUR_API_KEY"

# Requires the time.UnixDate format.
# On Linux: expiry=`date +'%a %b %d %H:%M:%S %Z %Y' --date '+6 hours'`
expiry="Mon Apr 14 19:06:45 EDT 2025"

echo "{
    \"headers\": {
        \"x-buildbuddy-api-key\": \"${apikey}\"
    },
    \"expiry\": \"${expiry}\"
}"
```

Example usage (after EngFlow Chromium setup):
```
$ cd /path/to/chromium/src
$ python3 third_party/reclient_configs/check_reclient_works.py --src_dir=.
Proxy started successfully.
Remotely executing "echo hello > hello"
Received expected result
Shutting down reproxy
RBE Stats: down 19 B, up 513 B, 1 remote execution
```

To facilitate debugging you can add additional environment variables:

```
export GODEBUG=http2debug=2
export GRPC_GO_LOG_VERBOSITY_LEVEL=99
export GRPC_GO_LOG_SEVERITY_LEVEL=info
```

And/or add additional command-line arguments in EngFlow's `check_reclient_works.py`:
```
diff --git check_reclient_works.py check_reclient_works.py
index c9fb764..079bb4e 100755
--- check_reclient_works.py
+++ check_reclient_works.py
@@ -58,6 +58,13 @@ def check_reclient_works():
     print('Starting reproxy')
     subprocess.check_call([
         f'{Paths.reclient_dir}/bootstrap',
+        '--log_dir',
+        '/Users/marshall/tmp/rewrapper',
+        '--proxy_log_dir',
+        '/Users/marshall/tmp/reproxy',
+        '--log_format=text',
+        '--logtostderr=true',
+        '--v=2',
         '--cfg',
         f'{Paths.reclient_cfgs_dir}/reproxy.cfg',
     ], env=env)
```

**Comment by magreenblatt** [comment]
[This](https://garden.tplant.com.au/Cloud/Building-Chromium) blog post suggests that setting `RBE_remote_headers=x-buildbuddy-api-key=mykey` may be a viable alternative to using a credentials_helper script.
